### PR TITLE
Feature branch PR4

### PR DESCRIPTION
**Pull request by mr-tms**

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+const express = require('express');
+const userRouter = require('./routes/userRoutes');
+
+const app = express();
+
+app.use(express.json());
+app.use('/api/v1/users', userRouter);
```
Comment (LiudasDEM) [inline app.js]: I'd suggest you to create completely separate router in routes directory:

`const router = express.Router();

router.use('/users', userRouter);

... app.js ...
const router = require('./routes')
app.use('/api/v1', router)
`

File: controllers/userController.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
+const User = require('../models/userModel');
+
+exports.getAllUsers = async (req, res) => {
+  const users = await User.find();
+
+  res.status(200).json({
+    status: 'success',
```
Comment (LiudasDEM) [inline controllers/userController.js]: 200 code already states that it is success, no need to add "status: 'success' property. Also, clean restapi for /GET endpoint should return array of items

File: controllers/userController.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    data: { users }
+  });
+};
+
+exports.createUser = async (req, res) => {
+  const newUser = await User.create(req.body);
+
+  res.status(201).json({
```
Comment (LiudasDEM) [inline controllers/userController.js]: 201 shouldnt return any body, instead it should add location header - where the new resource can be found. For example 

res.set('location', '/api/v1/users/${newUser.id}/`)

File: controllers/userController.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
+const User = require('../models/userModel');
+
+exports.getAllUsers = async (req, res) => {
+  const users = await User.find();
```
Comment (LiudasDEM) [inline controllers/userController.js]: consider implementing pagination

File: controllers/userController.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+  res.status(200).json({
+    status: 'success',
+    data: { users }
+  });
+};
+
+exports.createUser = async (req, res) => {
+  const newUser = await User.create(req.body);
```
Comment (LiudasDEM) [inline controllers/userController.js]: consider implementing req.body validation

**Review by LiudasDEM — COMMENTED** [review]
personally I'm no a fan of controller and router separation, I'd rather write:

router.get('users', function(req, res) {...});

but that's up to you :)

File: package.json [hunk 6]
```diff
@@ -2,10 +2,14 @@
...
+  "main": "server.ts",
   "scripts": {
-    "start": "nodemon index.js",
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "start": "nodemon src/server.ts",
+    "test": "npm test",
+    "build": "rimraf ./build && tsc",
+    "start:prod": "npm run build && node build/server.ts",
+    "lint": "eslint . --ext .ts",
+    "lint-and-fix": "eslint . --ext .ts --fix"
```
Comment (LiudasDEM) [inline package.json]: lint:fix name would be more consistent, but I'm not a fan of this commit at all - basically you allow a script to edit your code :/

File: src/controllers/userController.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+import User from '../models/userModel';
+
+export const getAllUsers = async (req: Request, res: Response): Promise<void> => {
+  const { page = 1, limit = 10 } = req.query;
+
+  try {
+    const users = await User.find()
+      .limit(Number(limit) * 1)
+      .skip((Number(page) - 1) * Number(limit))
+      .exec();
```
Comment (LiudasDEM) [inline src/controllers/userController.ts]: .exec(); is not necessary

File: src/controllers/userController.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    
+    const count = await User.countDocuments().exec();
+    
+    res.status(200).json({
```
Comment (LiudasDEM) [inline src/controllers/userController.ts]: when building rest-api always return only actual data, no meta data. You should do res.json(users) and that would be sufficient. If you want to sent how many pages there are (or even better - how many items in a collection) use headers pair:
Frontend sends: Return-Total-Count: 'true'
Backend returns: X-Total-Count: 'actual-count'

this is the cleanest communication

File: src/controllers/userController.ts [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    res.json(error.message);
+  }
+};
+
+export const getUser = async (req: Request, res: Response): Promise<void> => {
+  const userId = req.params.userId;
+
+  try {
+    const user = await User.findById(userId).exec();
+    res.status(200).json({
```
Comment (LiudasDEM) [inline src/controllers/userController.ts]: don't return an object with user param, simply json(user) is sufficient

File: src/controllers/userController.ts [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    res.status(400).json({ message: error.message });
+  }
+};
+
+export const createUser = async (req: Request, res: Response): Promise<void> => {
+  const newUser = await User.create(req.body);
+  
+  try {
+    res
+      .append('Content-Location', `/api/v1/users/${newUser._id}`)
```
Comment (LiudasDEM) [inline src/controllers/userController.ts]: 'Location' is the way to go

File: src/controllers/userController.ts [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    res.json(error.message);
+  }
+};
+
+export const deleteUser = async (req: Request, res: Response): Promise<void> => {
+  const userId = req.params.userId;
+  
+  try {
+    await User.findByIdAndDelete(userId).exec();
+    res.status(204).json({ message: 'User has been deleted.'});
```
Comment (LiudasDEM) [inline src/controllers/userController.ts]: 204 means no content, so res.sendStatus(204) is cleaner

**Review by LiudasDEM — CHANGES_REQUESTED** [review]

File: src/controllers/userController.ts (hunk 7 above)
Comment (mr-tms) [reply src/controllers/userController.ts]: Typescript issues warning regarding unnecessary await without .exec(). Didn't find another way to workaround this.

**Review by LiudasDEM — APPROVED** [review]
Looks good to me, but I also think we should schedule a meeting someday next week, I'd like to show you some things in "personal". The code seems to work completely fine, but I see some style improvements which you might agree on or completely disagree, that will be up to you

**Pull request merged**